File: apps/api/src/building-models/building-models.service.ts
import { HttpStatus, Injectable, Inject } from '@nestjs/common';
import { Readable } from 'node:stream';
import { pipeline } from 'node:stream/promises';
import { randomUUID } from 'node:crypto';
import { WS_EVENTS, BuildingModelVersionDto } from '@nodescope/shared';
import { NodeScopeException } from '../common/filters/global-exception.filter';
import { meteredHashingStream, UploadTooLargeError, InvalidIfcError } from './metered-hashing-stream';
import { OrgMemberContext } from '../organizations/org-context.types';
import { AuditService } from '../audit/audit.service';
import { IRealtimeService, REALTIME_SERVICE } from '../realtime/realtime.types';
import { PermissionsService } from '../permissions/permissions.service';
import { PropertiesService } from '../properties/properties.service';
import { StorageService } from '../storage/storage.service';
import { BuildingModelsRepository } from './building-models.repository';
import { toBuildingModelDto, toVersionDto } from './building-models.mapper';

/**
 * Orchestrates building-model reads, activation/rollback, and version deletion over
 * the Phase A repository + storage seam. Mutations are OWNER/ADMIN-gated at the
 * controller (`@OrgRoles`); audit + `v1:buildingModel:*` org-room events ride here.
 * Upload (Task 3) + download (Task 4) extend this service.
 */
@Injectable()
export class BuildingModelsService {
  constructor(
    private readonly repo: BuildingModelsRepository,
    private readonly storage: StorageService,
    private readonly properties: PropertiesService,
    private readonly audit: AuditService,
    private readonly permissions: PermissionsService,
    @Inject(REALTIME_SERVICE) private readonly realtime: IRealtimeService,
  ) {}

  private async loadModelOr404(member: OrgMemberContext, propertyId: string) {
    const model = await this.repo.findByProperty(member.organizationId, propertyId);
    if (!model) throw new NodeScopeException('MODEL_001', 'BUILDING_MODEL_NOT_FOUND', HttpStatus.NOT_FOUND);
    return model;
  }

  /**
   * F3 read gate. OWNER sees every building; otherwise the building (its own propertyId is the
   * governing site) must be in the member's assigned scope. Out-of-scope reads 404 (not 403) so a
   * model's existence is not revealed outside scope — mirrors the read-isolation other modules apply.
   */
  private async assertReadable(member: OrgMemberContext, propertyId: string): Promise<void> {
    if (member.role === 'OWNER') return;
    if (!(await this.permissions.inScope(member.organizationId, member.id, propertyId))) {
      throw new NodeScopeException('MODEL_001', 'BUILDING_MODEL_NOT_FOUND', HttpStatus.NOT_FOUND);
    }
  }

  async getModel(member: OrgMemberContext, propertyId: string) {
    await this.assertReadable(member, propertyId);
    return toBuildingModelDto(await this.loadModelOr404(member, propertyId));
  }

  async listVersions(member: OrgMemberContext, propertyId: string) {
    await this.assertReadable(member, propertyId);
    const model = await this.loadModelOr404(member, propertyId);
    return (await this.repo.listVersions(member.organizationId, model.id)).map(toVersionDto);
  }

  async activateVersion(member: OrgMemberContext, propertyId: string, versionId: string) {
    await this.permissions.assertCanConfigure(member, propertyId);
    const model = await this.loadModelOr404(member, propertyId);
    const version = await this.repo.findVersion(member.organizationId, versionId);
    if (!version || version.buildingModelId !== model.id) {
      throw new NodeScopeException('MODEL_004', 'MODEL_VERSION_NOT_FOUND', HttpStatus.NOT_FOUND);
    }
    const ok = await this.repo.setActiveVersion(member.organizationId, model.id, versionId, model.version);
    if (!ok) throw new NodeScopeException('SYNC_001', 'EDIT_CONFLICT', HttpStatus.CONFLICT);
    this.realtime.pushToOrg(member.organizationId, WS_EVENTS.BUILDING_MODEL_ACTIVATED, { propertyId, versionId });
    return toBuildingModelDto(await this.loadModelOr404(member, propertyId));
  }

  async deleteVersion(member: OrgMemberContext, propertyId: string, versionId: string): Promise<void> {
    await this.permissions.assertCanConfigure(member, propertyId);
    const model = await this.loadModelOr404(member, propertyId);
    const version = await this.repo.findVersion(member.organizationId, versionId);
    if (!version || version.buildingModelId !== model.id) {
      throw new NodeScopeException('MODEL_004', 'MODEL_VERSION_NOT_FOUND', HttpStatus.NOT_FOUND);
    }
    if (model.activeVersionId === versionId) {
      throw new NodeScopeException('MODEL_005', 'CANNOT_DELETE_ACTIVE_VERSION', HttpStatus.CONFLICT);
    }
    await this.storage.deleteObject(version.storageKey);
    await this.repo.deleteVersion(member.organizationId, versionId);
    await this.audit.recordDelete(member.organizationId, 'BuildingModelVersion', version);
    this.realtime.pushToOrg(member.organizationId, WS_EVENTS.BUILDING_MODEL_DELETED, { versionId });
  }

  async getActiveFile(member: OrgMemberContext, propertyId: string): Promise<{ stream: Readable; fileName: string; sizeBytes: number }> {
    await this.assertReadable(member, propertyId);
    const model = await this.loadModelOr404(member, propertyId);
    if (!model.activeVersionId) {
      throw new NodeScopeException('MODEL_004', 'MODEL_VERSION_NOT_FOUND', HttpStatus.NOT_FOUND);
    }
    const version = await this.repo.findVersion(member.organizationId, model.activeVersionId);
    if (!version) throw new NodeScopeException('MODEL_004', 'MODEL_VERSION_NOT_FOUND', HttpStatus.NOT_FOUND);
    return { stream: await this.storage.getObjectStream(version.storageKey), fileName: version.fileName, sizeBytes: version.sizeBytes };
  }

  async getVersionFile(
    member: OrgMemberContext,
    propertyId: string,
    versionId: string,
  ): Promise<{ stream: Readable; fileName: string; sizeBytes: number }> {
    await this.assertReadable(member, propertyId);
    const model = await this.loadModelOr404(member, propertyId);
    const version = await this.repo.findVersion(member.organizationId, versionId);
    if (!version || version.buildingModelId !== model.id) {
      throw new NodeScopeException('MODEL_004', 'MODEL_VERSION_NOT_FOUND', HttpStatus.NOT_FOUND);
    }
    return { stream: await this.storage.getObjectStream(version.storageKey), fileName: version.fileName, sizeBytes: version.sizeBytes };
  }

  /**
   * Proxied streaming upload: pipe the raw request body through the metered/hashing
   * transform into the bucket, then (only after the object lands — §11 orphan
   * mitigation) write the immutable version row and flip the active pointer.
   */
  async uploadVersion(
    member: OrgMemberContext,
    propertyId: string,
    fileName: string,
    units: string | null,
    body: Readable,
  ): Promise<BuildingModelVersionDto> {
    // F3 per-site configure scope FIRST — OWNER any; out-of-scope ADMIN → PERM_001; MEMBER → ORG_003
    // (the controller's @OrgRoles already blocks MEMBER, this adds the per-site authority). Gating
    // before any DB lookup means an out-of-scope ADMIN learns nothing about the target property's
    // existence/type, and never triggers the (expensive) streamed upload. Mirrors activate/delete.
    await this.permissions.assertCanConfigure(member, propertyId);

    const property = await this.properties.findInOrg(member.organizationId, propertyId);
    if (!property) throw new NodeScopeException('MODEL_001', 'BUILDING_MODEL_NOT_FOUND', HttpStatus.NOT_FOUND);
    if (property.type !== 'BUILDING') {
      throw new NodeScopeException('MODEL_002', 'PROPERTY_NOT_BUILDING', HttpStatus.UNPROCESSABLE_ENTITY);
    }

    const versionId = randomUUID();
    const key = this.storage.buildVersionKey(member.organizationId, propertyId, versionId);
    const maxBytes = parseInt(process.env.MODEL_MAX_BYTES ?? '209715200', 10);
    const { transform, result } = meteredHashingStream(maxBytes);
    // Capture the transform's validation error directly — the AWS Upload may wrap it.
    let streamError: Error | undefined;
    transform.on('error', (e: Error) => {
      streamError = e;
    });
    try {
      // pipeline propagates request abort/error into the transform. Run the storage consumer at
      // the same time so backpressure works and both promises settle on either-side failure.
      await Promise.all([
        pipeline(body, transform),
        this.storage.putObjectStream(key, transform, 'application/octet-stream'),
      ]);
    } catch (e) {
      await this.storage.deleteObject(key).catch(() => undefined); // clean any partial object
      const err = streamError ?? e;
      if (err instanceof UploadTooLargeError) {
        throw new NodeScopeException('MODEL_006', 'MODEL_FILE_TOO_LARGE', HttpStatus.PAYLOAD_TOO_LARGE);
      }
      if (err instanceof InvalidIfcError) {
        throw new NodeScopeException('MODEL_007', 'INVALID_IFC_FILE', HttpStatus.UNPROCESSABLE_ENTITY);
      }
      throw e;
    }
    const { contentHash, sizeBytes } = result();

    // Get-or-create the model only AFTER the object has landed — a failed upload
    // leaves no empty model. Name defaults to the building's name on first upload.
    let model;
    let versionNumber: number;
    let version;
    try {
      model =
        (await this.repo.findByProperty(member.organizationId, propertyId)) ??
        (await this.repo.createModel({ organizationId: member.organizationId, propertyId, name: property.name }));
      versionNumber = await this.repo.nextVersionNumber(member.organizationId, model.id);
      version = await this.repo.createVersion({
        organizationId: member.organizationId,
        buildingModelId: model.id,
        versionNumber,
        storageKey: key,
        fileName: fileName || 'model.ifc',
        contentHash,
        sizeBytes,
        units,
        uploadedByMemberId: member.id,
      });
    } catch (error) {
      // Until createVersion succeeds no DB row owns the landed blob.
      await this.storage.deleteObject(key).catch(() => undefined);
      throw error;
    }
    await this.repo.setActiveVersion(member.organizationId, model.id, version.id, model.version);
    await this.audit.recordCreate(member.organizationId, 'BuildingModelVersion', version);
    this.realtime.pushToOrg(member.organizationId, WS_EVENTS.BUILDING_MODEL_VERSION_UPLOADED, {
      propertyId,
      versionId: version.id,
      versionNumber,
    });
    return toVersionDto(version);
  }
}

File: src/Modules/Inventory/NodeScope.Modules.Inventory.Application/BuildingModels/BuildingModelsService.cs
using System.Security.Cryptography;
using System.Text.Json;
using NodeScope.Contracts.Realtime;
using NodeScope.Modules.Inventory.Application.Properties;
using NodeScope.Modules.Inventory.Domain;
using NodeScope.Platform.Abstractions;

namespace NodeScope.Modules.Inventory.Application.BuildingModels;

/// <summary>
/// Building-model reads, uploads, activation, and version deletion (Node's
/// <c>BuildingModelsService</c>). Mutations are OWNER/ADMIN-gated at the endpoint; this layer
/// adds the per-site F3 authority, the audit trail, and the <c>v1:buildingModel:*</c> events.
/// </summary>
public sealed class BuildingModelsService
{
    /// <summary>200 MiB, the Node default for <c>MODEL_MAX_BYTES</c>.</summary>
    public const long DefaultMaxBytes = 209_715_200;

    private const int MaxMetadataBytes = 33_554_432;

    private static readonly JsonSerializerOptions MetadataJson =
        new(JsonSerializerDefaults.Web);

    private readonly IBuildingModelRepository _models;
    private readonly IPropertyRepository _properties;
    private readonly IObjectStorage _storage;
    private readonly IPermissionScopeService _permissions;
    private readonly IRealtimeService _realtime;
    private readonly IAuditService _audit;

    public BuildingModelsService(
        IBuildingModelRepository models,
        IPropertyRepository properties,
        IObjectStorage storage,
        IPermissionScopeService permissions,
        IRealtimeService realtime,
        IAuditService audit)
    {
        _models = models;
        _properties = properties;
        _storage = storage;
        _permissions = permissions;
        _realtime = realtime;
        _audit = audit;
    }

    public async Task<BuildingModelDto> GetModelAsync(
        OrgMemberContext member,
        string propertyId,
        CancellationToken cancellationToken)
    {
        await AssertReadableAsync(member, propertyId, cancellationToken);
        var model = await RequireModelAsync(member.OrganizationId, propertyId, cancellationToken);
        return model.ToDto();
    }

    public async Task<IReadOnlyList<BuildingModelVersionDto>> ListVersionsAsync(
        OrgMemberContext member,
        string propertyId,
        CancellationToken cancellationToken)
    {
        await AssertReadableAsync(member, propertyId, cancellationToken);
        var model = await RequireModelAsync(member.OrganizationId, propertyId, cancellationToken);
        var versions = await _models.ListVersionsAsync(member.OrganizationId, model.Id, cancellationToken);
        return [.. versions.Select(version => version.ToDto())];
    }

    /// <summary>
    /// Streams the request body into storage, validating size and the IFC prefix as it flows,
    /// and only then writes the version row: a failed upload must leave neither an orphaned
    /// object nor an empty model.
    /// </summary>
    public async Task<BuildingModelVersionDto> UploadVersionAsync(
        OrgMemberContext member,
        string propertyId,
        string? fileName,
        string? units,
        Stream body,
        long maxBytes,
        bool activate,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(member);

        // The F3 gate runs before any lookup, so an out-of-scope ADMIN learns nothing about the
        // target property and never triggers the expensive upload.
        await _permissions.AssertCanConfigureAsync(member, propertyId, cancellationToken);

        var property = await _properties.FindAsync(member.OrganizationId, propertyId, null, cancellationToken)
            ?? throw InventoryErrors.BuildingModelNotFound();
        if (property.Type != PropertyType.Building)
        {
            throw InventoryErrors.PropertyNotBuilding();
        }

        var versionId = Guid.NewGuid().ToString();
        var key = StorageKeys.BuildingModelVersion(member.OrganizationId, propertyId, versionId);
        string contentHash;
        int sizeBytes;
        await using (var content = new IfcContentStream(body, maxBytes))
        {
            try
            {
                await _storage.PutAsync(key, content, "application/octet-stream", cancellationToken);
            }
            catch (Exception)
            {
                await DeleteQuietlyAsync(key);
                throw;
            }

            contentHash = content.ContentHash();
            sizeBytes = (int)content.SizeBytes;
        }

        BuildingModelRecord model;
        BuildingModelVersionRecord version;
        try
        {
            model = await _models.FindByPropertyAsync(member.OrganizationId, propertyId, cancellationToken)
                ?? await _models.CreateModelAsync(
                    member.OrganizationId, propertyId, property.Name, cancellationToken);
            var versionNumber = await _models.NextVersionNumberAsync(
                member.OrganizationId, model.Id, cancellationToken);
            version = await _models.CreateVersionAsync(
                new NewBuildingModelVersion(
                    versionId,
                    member.OrganizationId,
                    model.Id,
                    versionNumber,
                    key,
                    string.IsNullOrEmpty(fileName) ? "model.ifc" : fileName,
                    contentHash,
                    sizeBytes,
                    units,
                    member.MemberId),
                cancellationToken);
        }
        catch (Exception)
        {
            // Until the version row exists, nothing in the database owns the landed object.
            await DeleteQuietlyAsync(key);
            throw;
        }

        if (activate
            && !await _models.SetActiveVersionAsync(
                member.OrganizationId, model.Id, version.Id, model.Version, cancellationToken))
        {
            throw Changesets.EditConflict();
        }

        await _audit.RecordCreateAsync(
            member.OrganizationId, "BuildingModelVersion", version.Id, version.ToDto(), cancellationToken);
        await _realtime.PushToOrgAsync(
            member.OrganizationId,
            WsEvents.BuildingModelVersionUploaded,
            new { propertyId, versionId = version.Id, versionNumber = version.VersionNumber },
            cancellationToken);
        return version.ToDto();
    }

    /// <summary>
    /// Stores the cross-platform render artifact paired with an immutable IFC version.
    /// The PUT is intentionally idempotent: retrying after a lost response atomically
    /// replaces the same object key with the same derived bytes.
    /// </summary>
    public async Task<BuildingModelGeometryDto> UploadGeometryAsync(
        OrgMemberContext member,
        string propertyId,
        string versionId,
        Stream body,
        long maxBytes,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(member);
        await _permissions.AssertCanConfigureAsync(member, propertyId, cancellationToken);
        var model = await RequireModelAsync(member.OrganizationId, propertyId, cancellationToken);
        _ = await RequireVersionAsync(member.OrganizationId, model.Id, versionId, cancellationToken);

        var key = StorageKeys.BuildingModelGeometry(member.OrganizationId, propertyId, versionId);
        string contentHash;
        int sizeBytes;
        await using (var content = new WexBimContentStream(body, maxBytes))
        {
            await _storage.PutAsync(
                key, content, "application/vnd.xbim.wexbim", cancellationToken);
            contentHash = content.ContentHash();
            sizeBytes = checked((int)content.SizeBytes);
        }

        return new BuildingModelGeometryDto(versionId, "WEXBIM", contentHash, sizeBytes);
    }

    public async Task<BuildingModelMetadataReceiptDto> UploadMetadataAsync(
        OrgMemberContext member,
        string propertyId,
        string versionId,
        UploadBuildingModelMetadataRequest request,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(member);
        ArgumentNullException.ThrowIfNull(request);
        await _permissions.AssertCanConfigureAsync(member, propertyId, cancellationToken);
        var model = await RequireModelAsync(member.OrganizationId, propertyId, cancellationToken);
        _ = await RequireVersionAsync(member.OrganizationId, model.Id, versionId, cancellationToken);
        if (request.Validate().Count > 0 || request.Elements is not { } elements)
        {
            throw InventoryErrors.InvalidModelMetadata();
        }

        var document = new StoredBuildingModelMetadata(
            request.FormatVersion,
            elements);
        var content = JsonSerializer.SerializeToUtf8Bytes(document, MetadataJson);
        if (content.Length > MaxMetadataBytes)
        {
            throw InventoryErrors.InvalidModelMetadata();
        }

        var key = StorageKeys.BuildingModelMetadata(
            member.OrganizationId,
            propertyId,
            versionId);
        using var stream = new MemoryStream(content, writable: false);
        await _storage.PutAsync(
            key,
            stream,
            "application/vnd.nodescope.ifc-elements+json",
            cancellationToken);
        return new BuildingModelMetadataReceiptDto(
            versionId,
            request.FormatVersion,
            Convert.ToHexStringLower(SHA256.HashData(content)),
            content.Length,
            elements.Count);
    }

    public async Task<BuildingModelDto> ActivateVersionAsync(
        OrgMemberContext member,
        string propertyId,
        string versionId,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(member);
        await _permissions.AssertCanConfigureAsync(member, propertyId, cancellationToken);
        var model = await RequireModelAsync(member.OrganizationId, propertyId, cancellationToken);
        _ = await RequireVersionAsync(member.OrganizationId, model.Id, versionId, cancellationToken);

        if (!await _models.SetActiveVersionAsync(
            member.OrganizationId, model.Id, versionId, model.Version, cancellationToken))
        {
            throw Changesets.EditConflict();
        }

        await _realtime.PushToOrgAsync(
            member.OrganizationId,
            WsEvents.BuildingModelActivated,
            new { propertyId, versionId },
            cancellationToken);
        return (await RequireModelAsync(member.OrganizationId, propertyId, cancellationToken)).ToDto();
    }

    public async Task DeleteVersionAsync(
        OrgMemberContext member,
        string propertyId,
        string versionId,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(member);
        await _permissions.AssertCanConfigureAsync(member, propertyId, cancellationToken);
        var model = await RequireModelAsync(member.OrganizationId, propertyId, cancellationToken);
        var version = await RequireVersionAsync(member.OrganizationId, model.Id, versionId, cancellationToken);
        if (model.ActiveVersionId == versionId)
        {
            throw InventoryErrors.CannotDeleteActiveVersion();
        }

        // Remove the owning row before its objects. A storage outage may leave an
        // unreachable orphan for later garbage collection, but it must never leave
        // a live database version pointing at an IFC that was already deleted.
        await _models.DeleteVersionAsync(member.OrganizationId, versionId, cancellationToken);
        try
        {
            await _audit.RecordDeleteAsync(
                member.OrganizationId, "BuildingModelVersion", versionId, version.ToDto(), cancellationToken);
            await _realtime.PushToOrgAsync(
                member.OrganizationId, WsEvents.BuildingModelDeleted, new { versionId }, cancellationToken);
        }
        finally
        {
            await DeleteQuietlyAsync(version.StorageKey);
            await DeleteQuietlyAsync(
                StorageKeys.BuildingModelGeometry(member.OrganizationId, propertyId, versionId));
            await DeleteQuietlyAsync(
                StorageKeys.BuildingModelMetadata(member.OrganizationId, propertyId, versionId));
        }
    }

    public async Task<ModelFile> GetActiveFileAsync(
        OrgMemberContext member,
        string propertyId,
        CancellationToken cancellationToken)
    {
        await AssertReadableAsync(member, propertyId, cancellationToken);
        var model = await RequireModelAsync(member.OrganizationId, propertyId, cancellationToken);
        if (model.ActiveVersionId is null)
        {
            throw InventoryErrors.ModelVersionNotFound();
        }

        var version = await _models.FindVersionAsync(member.OrganizationId, model.ActiveVersionId, cancellationToken)
            ?? throw InventoryErrors.ModelVersionNotFound();
        return await OpenAsync(version, cancellationToken);
    }

    public async Task<ModelFile> GetActiveGeometryAsync(
        OrgMemberContext member,
        string propertyId,
        CancellationToken cancellationToken)
    {
        await AssertReadableAsync(member, propertyId, cancellationToken);
        var model = await RequireModelAsync(member.OrganizationId, propertyId, cancellationToken);
        if (model.ActiveVersionId is null)
        {
            throw InventoryErrors.ModelVersionNotFound();
        }

        var version = await _models.FindVersionAsync(
            member.OrganizationId, model.ActiveVersionId, cancellationToken)
            ?? throw InventoryErrors.ModelVersionNotFound();
        return await OpenGeometryAsync(member.OrganizationId, propertyId, version, cancellationToken);
    }

    public async Task<ModelFile> GetVersionFileAsync(
        OrgMemberContext member,
        string propertyId,
        string versionId,
        CancellationToken cancellationToken)
    {
        await AssertReadableAsync(member, propertyId, cancellationToken);
        var model = await RequireModelAsync(member.OrganizationId, propertyId, cancellationToken);
        var version = await RequireVersionAsync(member.OrganizationId, model.Id, versionId, cancellationToken);
        return await OpenAsync(version, cancellationToken);
    }

    public async Task<ModelFile> GetVersionGeometryAsync(
        OrgMemberContext member,
        string propertyId,
        string versionId,
        CancellationToken cancellationToken)
    {
        await AssertReadableAsync(member, propertyId, cancellationToken);
        var model = await RequireModelAsync(member.OrganizationId, propertyId, cancellationToken);
        var version = await RequireVersionAsync(member.OrganizationId, model.Id, versionId, cancellationToken);
        return await OpenGeometryAsync(member.OrganizationId, propertyId, version, cancellationToken);
    }

    public async Task<BuildingModelMetadataDto> GetActiveMetadataAsync(
        OrgMemberContext member,
        string propertyId,
        CancellationToken cancellationToken)
    {
        await AssertReadableAsync(member, propertyId, cancellationToken);
        var model = await RequireModelAsync(member.OrganizationId, propertyId, cancellationToken);
        if (model.ActiveVersionId is null)
        {
            throw InventoryErrors.ModelVersionNotFound();
        }

        var version = await _models.FindVersionAsync(
            member.OrganizationId,
            model.ActiveVersionId,
            cancellationToken)
            ?? throw InventoryErrors.ModelVersionNotFound();
        return await ReadMetadataAsync(
            member.OrganizationId,
            propertyId,
            version.Id,
            cancellationToken);
    }

    public async Task<BuildingModelMetadataDto> GetVersionMetadataAsync(
        OrgMemberContext member,
        string propertyId,
        string versionId,
        CancellationToken cancellationToken)
    {
        await AssertReadableAsync(member, propertyId, cancellationToken);
        var model = await RequireModelAsync(member.OrganizationId, propertyId, cancellationToken);
        _ = await RequireVersionAsync(member.OrganizationId, model.Id, versionId, cancellationToken);
        return await ReadMetadataAsync(
            member.OrganizationId,
            propertyId,
            versionId,
            cancellationToken);
    }

    /// <summary>
    /// Read gate: the building is its own governing site. An out-of-scope read 404s rather than
    /// 403s, so a model's existence never leaks outside the member's scope.
    /// </summary>
    private async Task AssertReadableAsync(
        OrgMemberContext member,
        string propertyId,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(member);
        if (!await _permissions.IsInScopeAsync(member, propertyId, cancellationToken))
        {
            throw InventoryErrors.BuildingModelNotFound();
        }
    }

    private async Task<BuildingModelRecord> RequireModelAsync(
        string organizationId,
        string propertyId,
        CancellationToken cancellationToken) =>
        await _models.FindByPropertyAsync(organizationId, propertyId, cancellationToken)
        ?? throw InventoryErrors.BuildingModelNotFound();

    private async Task<BuildingModelVersionRecord> RequireVersionAsync(
        string organizationId,
        string modelId,
        string versionId,
        CancellationToken cancellationToken)
    {
        var version = await _models.FindVersionAsync(organizationId, versionId, cancellationToken);
        return version is null || version.BuildingModelId != modelId
            ? throw InventoryErrors.ModelVersionNotFound()
            : version;
    }

    private async Task<ModelFile> OpenAsync(BuildingModelVersionRecord version, CancellationToken cancellationToken) =>
        new(await _storage.GetAsync(version.StorageKey, cancellationToken), version.FileName, version.SizeBytes);

    private async Task<ModelFile> OpenGeometryAsync(
        string organizationId,
        string propertyId,
        BuildingModelVersionRecord version,
        CancellationToken cancellationToken)
    {
        var key = StorageKeys.BuildingModelGeometry(organizationId, propertyId, version.Id);
        if (!await _storage.ExistsAsync(key, cancellationToken))
        {
            throw InventoryErrors.ModelGeometryNotFound();
        }

        var fileName = $"{Path.GetFileNameWithoutExtension(version.FileName)}.wexbim";
        return new ModelFile(await _storage.GetAsync(key, cancellationToken), fileName, 0);
    }

    private async Task<BuildingModelMetadataDto> ReadMetadataAsync(
        string organizationId,
        string propertyId,
        string versionId,
        CancellationToken cancellationToken)
    {
        var key = StorageKeys.BuildingModelMetadata(organizationId, propertyId, versionId);
        if (!await _storage.ExistsAsync(key, cancellationToken))
        {
            throw InventoryErrors.ModelMetadataNotFound();
        }

        await using var source = await _storage.GetAsync(key, cancellationToken);
        using var content = new MemoryStream();
        var buffer = new byte[81920];
        while (true)
        {
            var read = await source.ReadAsync(buffer, cancellationToken);
            if (read == 0)
            {
                break;
            }

            if (content.Length > MaxMetadataBytes - read)
            {
                throw InventoryErrors.InvalidModelMetadata();
            }

            await content.WriteAsync(buffer.AsMemory(0, read), cancellationToken);
        }

        StoredBuildingModelMetadata? document;
        try
        {
            document = JsonSerializer.Deserialize<StoredBuildingModelMetadata>(
                content.GetBuffer().AsSpan(0, checked((int)content.Length)),
                MetadataJson);
        }
        catch (JsonException)
        {
            throw InventoryErrors.InvalidModelMetadata();
        }

        if (document is null || document.Elements is null)
        {
            throw InventoryErrors.InvalidModelMetadata();
        }

        var validation = new UploadBuildingModelMetadataRequest
        {
            FormatVersion = document.FormatVersion,
            Elements = document.Elements,
        };
        if (validation.Validate().Count > 0)
        {
            throw InventoryErrors.InvalidModelMetadata();
        }

        return new BuildingModelMetadataDto(
            versionId,
            document.FormatVersion,
            document.Elements);
    }

    /// <summary>
    /// Cleanup on a failed upload. Every failure is swallowed on purpose: this runs while an
    /// exception is already in flight, and a storage error here would replace the real cause
    /// with a misleading one. The worst case is one orphaned object.
    /// </summary>
    private async Task DeleteQuietlyAsync(string key)
    {
#pragma warning disable CA1031
        using var timeout = new CancellationTokenSource(TimeSpan.FromSeconds(30));
        try
        {
            await _storage.DeleteAsync(key, timeout.Token);
        }
        catch (Exception)
        {
        }
#pragma warning restore CA1031
    }

    private sealed record StoredBuildingModelMetadata(
        int FormatVersion,
        IReadOnlyList<BuildingModelElementDto> Elements);
}
